Removed finalize from NetworkSession

lutovich opened the pull request:
It contained logic for closing opened underlying connection and logging info about the leak. Closing of resources in finalize is not reliable and overridden finalize can hurt when many sessions are garbage collected.

Commit also introduces a system property `org.neo4j.driver.logSessionLeaks` to enable logging of unclosed/leaked sessions and log stacktrace of where they were instantiated. This is meant to help with session leak investigations. So finalization overhead will be present only when problem is present and is being investigated.

lutovich committed:
Removed finalize from NetworkSession

It contained logic for closing opened underlying connection and logging info
about the leak. Closing of resources in finalize is not reliable and overridden
finalize can hurt when many sessions are garbage collected.

Commit also introduces a system property `org.neo4j.driver.logSessionLeaks`
to enable logging of unclosed/leaked sessions and log stacktrace of where they
were instantiated. This is meant to help with session leak investigations. So
finalization overhead will be present only when problem is present and is
being investigated.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/LeakLoggingNetworkSession.java:
```diff
@@ -0,0 +1,65 @@
+    {
+        if ( isOpen() )
+        {
+            log.error( "Neo4j Session object leaked, please ensure that your application" +
```
error? or info? or warn?
These three are all printed out by default in command line logging.

lutovich replied:
Leak is basically an error, that is why I use `log.error` here.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/LeakLoggingNetworkSessionFactory.java:
```diff
@@ -0,0 +1,49 @@
+        this.logger = logging.getLog( LOGGER_NAME );
+    }
+
+    public static boolean leakLoggingEnabled()
```
wondering if we should enable it by a system variable or by a config option?
Such as
```Driver driver = Graphdatabase.Driver(uri, auth, Config.build().withSessionLeakLoggingEnabled().toConfig())```

lutovich replied:
My main motivation for using system property is that users would not need to recompile their apps for leak investigation. So it would be possible to just restart it with `-Dorg.neo4j.driver.logSessionLeaks=true`. Do you think this makes sense?

zhenlineo replied:
aha, that's good thinking. I was just thinking that they need to restart anyway, but yeah, if they do not need to change the code, that's even better

zhenlineo commented:
You need to update your file header

zhenlineo commented:
@lutovich Sorry, need to rebase. I am general +1 on this, while I am wondering if we want to add this special parameter as a system property or we should consider to add in Config and build a xml config reader

lutovich committed:
Move leaked session switch to config

It was previously in system property. This commit makes it part of `Config` so
we have a single place with all configuration parameters.
